### DISCUSSION TITLE
Implement AD Authentication in Perpective App

### DISCUSSION THREAD

**Discussion by akash-goel** [perspective-dev/perspective#1994] [Q&A] [answered]
Hi Team,

we want to implemented Security in Perspective App and want to secure the websocket connection in Tornado Perspective Application, On the Javscript we have noticed perpective.websocket(url) is not supporting additional headers and cookies , Can you please suggest how can we implement Authentication and Authorization in Perspective Tornado Application.

Thanks

**Comment by GREEB** [comment]
> The WebSocket protocol doesn’t handle authorization or authentication. Practically, this means that a WebSocket opened from a page behind auth doesn’t "automatically" receive any sort of auth. You need to take steps to also secure the WebSocket connection.
> 
> Since you cannot customize WebSocket headers from JavaScript, you’re limited to the "implicit" auth (i.e. Basic or cookies) that’s sent from the browser. The more common approach to generates a token from your normal HTTP server and then have the client send the token (either as a query string in the WebSocket path or as the first WebSocket message). The WebSocket server then validates that the token is valid.

[from the fastapi jwt docs with examples](https://indominusbyte.github.io/fastapi-jwt-auth/advanced-usage/websocket/)

**Comment by timkpaine** [comment, marked as answer]
Perspective only provides helpers for interacting over tornado/fastapi/aiohttp in python, you can put whatever authentication mechanisms you like in front of it.
